packages/remix-server-runtime/routes.ts

```diff
@@ -40,27 +40,50 @@ export interface ServerRoute extends Route {
   module: ServerRouteModule;
 }
 
+function groupRoutesByParentId(manifest: ServerRouteManifest) {
+  let routes: Record<string, Omit<ServerRoute, "children">[]> = {};
+
+  Object.values(manifest).forEach((route) => {
+    let parentId = route.parentId || "";
+    if (!routes[parentId]) {
+      routes[parentId] = [];
+    }
+    routes[parentId].push(route);
+  });
+
+  return routes;
+}
+
 export function createRoutes(
   manifest: ServerRouteManifest,
-  parentId?: string
+  parentId?: string,
+  routesByParentId?: Record<string, Omit<ServerRoute, "children">[]>
 ): ServerRoute[] {
-  return Object.entries(manifest)
-    .filter(([, route]) => route.parentId === parentId)
-    .map(([id, route]) => ({
+  // Create a map of routes by parentId to use recursively instead of
+  // repeatedly filtering the manifest.
+  routesByParentId ||= groupRoutesByParentId(manifest);
```

Comment:
@machour - would the [logical or assignment ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment)(|=) be transpiled correctly for older browsers in Remix's build process?

I recall recently that null coalescing operator was removed:
https://github.com/remix-run/remix/pull/4561

Reply:
As far as I know, this code will only run on the server, so it shouldn't cause any problem with the browser

Reply:
We ran into an issue on node with this on node 14 that should be fixed since we changed the target in our `babelrc`, but that might be worth confirming.

Reply:
I ended up removing the `||=` in [df6f364](https://github.com/remix-run/remix/pull/4748/commits/df6f364d92b80eef7747d188a31b99be895b3323) and switched it to use a default param value